### ISSUE TITLE
[APEX] SOQL Injection suffers scope leakage tracking variables.

### ISSUE THREAD

**Issue by rcorfieldffdc** [pmd/pmd#4650]
**Affects PMD Version:** 

7.0.0-RC3 (Erroneous code is apparent in the master branch at time of writing)

**Rule:**

APEXSOQLInjection
https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexsoqlinjection

**Description:**

The rule as coded keeps a HashSet of trusted variable names within the whole class. The variable is entered into the set if it appears safe. Assignment from string literal is safe and the variable remains marked safe if mutated and also into other scopes.

**Code Sample demonstrating the issue:**

```java

    public static void testMutationOfLiteral(String untrusted) {
        String aValue = 'Hello';
        aValue = aValue + untrusted;
        Database.query('SELECT Name From Account WHERE Name = ' + aValue);
    }

    public static void testLeakageBetweenScopes(String aValue) {
        Database.query('SELECT Name From Account WHERE Name = ' + aValue);
    }

```

**Expected outcome:**

There should be two violations here.

**Running PMD through:** *VSCode Integration*

**Aside**

I already have a pull request to look at another problem in this rule (#4646).
That has taken be down this rabbit hole as I try to fix a unit test I put into that branch concerning member variables. I ought to cut off that pull request to keep it small (backing out its failing test) so that this much larger change can be carried out in a new branch.

**Comment by rcorfieldffdc** [comment]
I have started work, but need to park it for a short while.

https://github.com/rcorfieldffdc/pmd/tree/defect/4650-apex-soql-injection-variable-handling-issues

**Comment by rcorfieldffdc** [comment]
TODO:

I've started trying to rewrite this using the existing AST, which is quite a job and will never be completely accurate. I notice that the Unused Method rule uses @nawforce multi file analyser. This has a documented API! Could it provide both an easier to analyse structure, and ability to chase data access outside the file? 

What would the cost be in terms of memory and performance for something that's meant to be a quick scan? We can do quite well treating the file as a boundary for quick scans. After all, this is a guide and the user can mark false positives if they wish. The work I've started would treat any incoming data or data that is mutable outside the file as untrusted. Sadly this would mean that things like ApexLib's `QueryBuilder` could never be trusted to return safe SOQL. Neither could base class methods in their Selectors. It is a trade-off.
